scene/main/viewport.cpp

```diff
@@ -4492,12 +4496,29 @@ void Viewport::set_use_xr(bool p_use_xr) {
 			RID rt = RS::get_singleton()->viewport_get_render_target(viewport);
 			RSG::texture_storage->render_target_set_override(rt, RID(), RID(), RID());
 		}
+
+		texture_rid = RenderingServer::get_singleton()->viewport_get_texture(viewport);
+		// Will need a new RID (i.e. switching from 2D render target to 3D buffer).
+		for (ViewportTexture *E : viewport_textures) {
+			RS::get_singleton()->texture_proxy_update(E->get_rid(), texture_rid);
+			E->emit_changed();
+		}
 	}
 }
 
+Viewport::ViewportMode Viewport::get_viewport_mode() const {
+	ERR_READ_THREAD_GUARD_V(VIEWPORT_MODE_2D_AND_3D);
+	return viewport_mode;
+}
+
+void Viewport::set_use_xr(bool p_use_xr) {
+	ERR_MAIN_THREAD_GUARD;
+	set_viewport_mode(p_use_xr ? VIEWPORT_MODE_XR : VIEWPORT_MODE_2D_AND_3D);
```

Comment:
I think one change you need to do here is that if `p_use_xr = false`, and `viewport_mode != VIEWPORT_MODE_XR`, that you skip this. There is a high risk here that `viewport_mode` is set to the required mode, but `set_use_xr` is still called afterwards (loading properties for instance), and we end up unsetting the mode we want.